File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.js
function possible(num, diffs, times, limit) {
  var timeTaken = times[0];
  var prevTime = times[0];
  for (let i = 1; i < diffs.length; i++) {
    var currentDiff = diffs[i];
    var currentTime = times[i];
    if (currentDiff <= num) {
      timeTaken += currentTime;
      prevTime = currentTime;
    } else {
      timeTaken += (prevTime + currentTime) * (currentDiff - num) + currentTime;
      prevTime = currentTime;
    }
    if (timeTaken > limit) {
      return false;
    }
  }
  return true;
}

function solution(diffs, times, limit) {
  var left = 1;  // Changed from 0 to 1
  var right = 300000;  // Changed to maximum difficulty

  while (left <= right) {
    var mid = Math.floor((left + right) / 2);
    if (possible(mid, diffs, times, limit)) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return left;  // Return left instead of answer
}

File: 프로그래머스/2/340212. ［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지/［PCCP 기출문제］ 2번 ／ 퍼즐 게임 챌린지.java
class Solution {
  int[] diffs, times;
  long limit;
  public int solution(int[] diffs, int[] times, long limit) {

    int answer = Integer.MAX_VALUE;
    this.diffs = diffs;
    this.times = times;
    this.limit = limit;

    int left = 1;
    int right = 1;
    for(int i : diffs){
      right = Math.max(i, right);
    }

    while(left <= right){
      int mid = (left + right) / 2;

      if(isPossible(mid)){
        answer = Math.min(mid, answer);
        right = mid - 1;
      }else {
        left = mid + 1;
      }
    }

    return answer;
  }

boolean isPossible(int level) {
    long totalTime = 0;
    long prevTime = 0;

    for (int i = 0; i < diffs.length; i++) {
        long diff = diffs[i];
        long time = times[i];

        long wrongs = Math.max(0, diff - level);
        totalTime += wrongs * (prevTime + time) + time;

        if (totalTime > limit) return false;

        prevTime = time;
    }

    return true;
}
}